### PR TITLE
JWT generation in Auth package

### PR DESCRIPTION
**Pull request by zeevallin**
Since we're checking on the service level for token validity, there's an argument to be made about providing tooling for easily generating tokens locally. This pull request should not break backward compatibility with [`v3.3.0`](https://github.com/LUSHDigital/microservice-core-golang/tree/3.3.0).

### CODE REVIEW

File: auth/tokenizer.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+// Tokeniser is the auth tokeniser for JSON Web Tokens
+type Tokeniser struct {
+	privateKey *rsa.PrivateKey
+	publicKey  *rsa.PublicKey
+	authIssuer string
+}
+
+// NewMockTokeniser creates a new tokeniser with a random key pair
+func NewMockTokeniser() (*Tokeniser, error) {
```
Comment (tobbbles) [inline auth/tokenizer.go]: I'd ideally like this to wrap `NewTokeniser`

**Review by nicklanng — APPROVED** [review]

File: auth/tokenizer.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+func NewPublicTokeniser(bPublicKey string) (*Tokeniser, error) {
+	publicKey, err := jwt.ParseRSAPublicKeyFromPEM([]byte(bPublicKey))
+	if err != nil {
+		return nil, err
+	}
+	return &Tokeniser{publicKey: publicKey}, nil
+}
+
+// NewTokeniser returns a new JWT instance
+func NewTokeniser(bPrivateKey, bPublicKey, issuer string) (*Tokeniser, error) {
```
Comment (tobbbles) [inline auth/tokenizer.go]: I'd prefer to see `bPrivateKey` and `bPublicKey` to be byte slices, to play nicely with `crypto/x509`.

Or even one better, `*rsa.PublicKey` and `*rsa.PrivateKey`

File: auth/tokenizer.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"os"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/pborman/uuid"
+	"gitlab.com/LUSHDigital/soa/first-class/api-gateway/service/config"
```
Comment (nicklanng) [inline auth/tokenizer.go]: this one needs updating

File: auth/tokenizer_test.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
+package auth_test
+
+import (
+	"log"
+	"testing"
+
+	"github.com/LUSHDigital/microservice-core-golang/auth"
+	"gitlab.platformserviceaccount.com/lush-soa/service/markets/service/test"
```
Comment (nicklanng) [inline auth/tokenizer_test.go]: this probably needs changing too!

File: auth/tokenizer.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		StandardClaims: jwt.StandardClaims{
+			ExpiresAt: time.Now().Add(time.Duration(config.TokenValidPeriod) * time.Minute).Unix(),
+			Issuer:    t.authIssuer,
+			Id:        uuid.New(),
+		},
+	}
+	// Create the token
+	newToken := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+	// Sign the token
+	token, err = newToken.SignedString(t.privateKey)
```
Comment (tobbbles) [inline auth/tokenizer.go]: Style nit, but drop the named returns and just `return newToken.SignedString(t.privateKey)`

**Review by tobbbles — CHANGES_REQUESTED** [review]
I'm happy with these changes! Just three code style points I've addressed

File: auth/errors_test.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package auth
+
+import (
+	"testing"
+
+	"gitlab.platformserviceaccount.com/lush-soa/service/markets/service/test"
```
Comment (nicklanng) [inline auth/errors_test.go]: here's one!

File: auth/consumer_test.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package auth_test
+
+import (
+	"testing"
+
+	"github.com/LUSHDigital/microservice-core-golang/auth"
+	"gitlab.platformserviceaccount.com/lush-soa/service/markets/service/test"
```
Comment (nicklanng) [inline auth/consumer_test.go]: here's a cheeky one!

File: auth/tokenizer.go (hunk 2 above)
Comment (nicklanng) [reply auth/tokenizer.go]: I think this is just a matter of when the conversion happens. They will come in as strings from the env vars and get turned into `[]byte`s and the `*rsa.things`. It just depends when in the method chain that conversion happens.

File: auth/auth_test.go [hunk 8]
```diff
@@ -39,9 +40,15 @@ W+kIFfkbaZVWbkUYAwIDAQAB
 )
 
 func TestMain(m *testing.M) {
-	if tokeniser, err = auth.NewTokeniser(testPrivateKey, testPublicKey, "testing"); err != nil {
+	if tokeniser, err = auth.NewTokaniserFromKeyPair([]byte(testPrivateKey), []byte(testPublicKey), "testing"); err != nil {
```
Comment (nicklanng) [inline auth/auth_test.go]: Tokanizer->Tokenizer

File: auth/tokenizer.go [hunk 9]
```diff
@@ -27,60 +31,61 @@ func NewMockTokeniser() (*Tokeniser, error) {
...
-		authIssuer: name,
-	}, nil
+	return NewTokeniser(privateKey, publicKey, issuer), nil
 }
 
-// NewPublicTokeniser returns a new JWT instance
-func NewPublicTokeniser(bPublicKey string) (*Tokeniser, error) {
-	publicKey, err := jwt.ParseRSAPublicKeyFromPEM([]byte(bPublicKey))
+// NewTokaniserFromPublicKey parses a public key to
+func NewTokaniserFromPublicKey(pkb []byte) (*Tokeniser, error) {
```
Comment (nicklanng) [inline auth/tokenizer.go]: Tokanizer -> Tokenizer

**Review by tobbbles — APPROVED** [review]

**Pull request merged**